File: 5.cpp
#include <iostream>
using namespace std;

class Solution {
public:
    string longestPalindrome(string s) {
        bool dp[s.length()][s.length()];
        for (int i = 0; i < s.length(); ++i) {
            for (int j = 0; j < s.length(); ++j) {
                dp[i][j] = false;
            }
        }
        int max_value = 0, r_s, r_e;
        for (int j = 0; j < s.length(); ++j) {
            for (int i = 0; i <= j; ++i) {
                if (i == j) {
                    dp[i][j] = true;
                }
                else if (i == j - 1) {
                    if (s[i] == s[j])
                        dp[i][j] = true;
                }
                else {
                    dp[i][j] = dp[i + 1][j - 1] && s[i] == s[j];
                }
                if (dp[i][j] && j - i + 1 > max_value) {
                    max_value = j - i + 1;
                    r_s = i; r_e = j + 1;
                }
            }
        }
        return s.substr(r_s, r_e - r_s);
    }
};



int main(void) {
    Solution s;
    cout << s.longestPalindrome("ABABCCC");
}

File: 5.java
class Solution {
    public String longestPalindrome(String s) {
        boolean[][] dp = new boolean[s.length()][s.length()];
        for (int i = 0; i < s.length(); ++i) {
            dp[i][i] = true;
        }
        int maxLength = 0;
        String retStr = s.substring(0, 1);
        for (int i = s.length() - 1; i >= 0; --i) {
            for (int j = i + 1; j < s.length(); ++j) {
                dp[i][j] = s.charAt(i) == s.charAt(j) && (dp[i + 1][j - 1] || i == j - 1);
                if (dp[i][j] && (j - i + 1) > maxLength) {
                    maxLength = j - i + 1;
                    retStr = s.substring(i, j + 1);
                }
            }
        }
        return retStr;
    }
}